# `Process.exist?` を作る

## 課題

以下のRubyプログラムと同様の処理を行う `Process.exist?` 組み込みメソッドを作成する

```ruby
def Process.exist?(pid)
  Process.kill(0, pid.to_i)
  true
rescue Errno::ESRCH # No such process
  false
rescue Errno::EPERM # Permission denied
  true
end
```

## 達成条件

下記の「やりかた」で追加したテストが通る。

## やりかた

 1. `Process`がどこで定義されているか探します。
 2. その付近に `proc_exist` 関数を実装し、`Process` に紐付けます。

    ```c
    rb_define_singleton_method(rb_mProcess, "exist?", proc_exist, 1);
    ```
    
 4. `make` が正常に終わるようにします。
 5. ソースコードの `test/ruby/test_process.rb` を開きます。`class TestProcess` の中に以下のようなテストを追加します。

    ```ruby
    def test_process_exist
      pid = spawn(RUBY, in: :close) # create a nonexistent PID
      Process.wait pid

      assert_equal(Process.exist?(pid), false)
      assert_equal(Process.exist?(Process.pid), true)
    end
    ```

 6. `make test-all` が正常に終わるようにします。

## ヒント

- `proc_exist` はRubyレベルで1引数なので、それとselfを加えてCでは2引数です。つまり以下のような感じです。

    ```C
    static VALUE proc_exist(VALUE self, VALUE pid);
    ```

- `Process.kill` を呼ぼうとすると大変なので、C関数の`kill`を使うと簡単。
- 上記の通りC関数の`kill`を使うのであればRubyのオブジェクトをCの型`pid_t`に変換する必要がありますが、それは`NUM2PIDT`を使います。

    ```C
    pid_t p = NUM2PIDT(pid);
    ```
    
- このメソッドは `true` や `false` を返す必要がありますが、それらはCのグローバル変数`Qtrue`と`Qfalse`として利用可能です。

    ```C
    return Qtrue;
    ```

- もし`Process.kill`を呼ぶのであれば、それは `rb_funcall` で実現可能です。

    ```C
	VALUE rb_mProcess = rb_const_get(rb_cObject, rb_intern("Process"));
    rb_funcall(rb_mProcess, rb_intern("kill"), 2, INT2FIX(0), arg);
    ```

- しかしながら、このメソッドは例外を投げるので、例外を受け取るように書かないといけません。それには `rb_rescue` を使います。

    ```C
    static VALUE
    proc_exist(VALUE self, VALUE pid)
    {
        return rb_rescue(proc_exist_begin, pid, proc_exist_rescue, self);
    }
    ```

    つまりこの場合はRubyで書いた時に `begin` に相当する部分と `rescue` に相当する部分をべつべつの関数に分けて作る必要があります。

- 上記の `rb_funcall` を使ってどうにかするというのは`begin`に相当する側の関数でやるやつです。
- `rescue`に相当する側の関数では、第2引数として渡ってくる例外オブジェクトを調べて、これに従って戻り値を作ります。

    ```C
	static VALUE proc_exist_rescue(VALUE self, VALUE exc);
	```
